baybe/constraints/base.py

```diff
@@ -103,6 +104,17 @@ def get_invalid(self, data: pd.DataFrame) -> pd.Index:
             The dataframe indices of rows where the constraint is violated.
         """
 
+    @abstractmethod
+    def get_polars_expression(self, df: pl.LazyFrame) -> pl.Expr:
```

Comment:
I suggest that not all constraints should have this method, so it should be removed as abstract method in the base class

This would enable you to apply polars filtering for all constraints where `hasattr(c, 'get_polars_expression')` is True and oldschcool pandas filtering for all others (mentioned this suggestion in the meeting)

this means we would not need all the "empty" implementations that return `pl.lit(True)`

Reply:
Polars product takes longer than Pandas if we can't filter out rows. 
So I'm going to use polars Cartesian product for for all constraints where `hasattr(c, 'get_polars_expression')` and Pandas for others in `from_product`.

Reply:
Update: I tried your suggested solution it made the code way more sophisticated and added to mypy typing errors.
I think since `pl.lit(True)` has no side effects, and maybe in the future you'll find some elegant ways to convert other constraints to Polars, keeping 'get_polars_expression' in all constraints mitigate problems.

Reply:
thats rather surprising, the suggestion would probably remove large chuks of code (the unnecessary `get_polars_expression`) while adding a few lines to the place where they are applied so how does it make it more sophisticated?

Reply:
I'm with @Scienfitz here: I can also not see how the code would become more complex... 🤔 !? Can you explain?

Reply:
This is more of a personal preference to avoid narrowing down type checking and have redundant checks that may complicate future changes.